kcidb/monitor/output.py

```diff
@@ -136,9 +138,26 @@ def render(self):
             ready to send, but missing the From header.
         """
         email = EmailMessage()
-        email["Subject"] = self.message.summary + self.obj.summarize()
+        subject = self.message.summary + self.obj.summarize()
+        email["Subject"] = subject
         email["To"] = ", ".join(self.message.recipients)
         email["X-KCIDB-Notification-ID"] = self.id
         email["X-KCIDB-Notification-Message-ID"] = self.message.id
-        email.set_content(self.message.description + self.obj.describe())
+        message = self.message.description + self.obj.describe()
+        email.set_content(message)
+        escaped_subject = html.escape(subject)
+        escaped_message = html.escape(message)
+        urls = re.compile(r'((http|https|git|ftp)?://[^|\s]+)')
+        escaped_message = urls.sub('<a href= "\\1">\\1</a>', escaped_message)
```

Comment:
This looks about right, however:
* There's not much sense to compile a regular expression separately, if we're not going to reuse the compiled regex, so could you just pass it to re.sub directly?
* Why did you put `?` after the `(http|https|git|ftp)`? URL wouldn't be valid without a schema, wouldn't it?
* Why is there `|` inside `[^|\s]` specifically? What case are you trying to handle?
* The [`html.escape()`](https://docs.python.org/3/library/html.html#html.escape) function does not escape quotes by default. Therefore it is not safe to put our substitution into the `href` attribute value as is. One easy fix, I think, could be to pass `quote = True` to both `html.escape()` invocations and simply let quotes be escaped everywhere.
* I think you can avoid the outside parentheses in the pattern, and substitute the whole match with `\g<0>` instead, but the latter is quite obscure, so I'm not sure if that would be better :smile:

Reply:
The | in [^|\s] is a typo.. Its [^\s]